File: packages/sdk/test/frontmatter.test.ts
import { strict as assert } from 'node:assert';
import { test } from 'node:test';
import type { ScannedProfile } from '../dist/index.js';
import { parseFrontmatter } from '../dist/lib/frontmatter.js';
import { resolveAgentProfile } from '../dist/lib/layer.js';
import { schemaErrors } from '../dist/lib/schemas.js';

test('document without frontmatter passes through', () => {
   const fm = parseFrontmatter('# Title\n\nBody.\n');
   assert.equal(fm.data, null);
   assert.equal(fm.error, undefined);
});

test('unterminated frontmatter block is an error', () => {
   const fm = parseFrontmatter('---\nid: x\n# never closed\n');
   assert.match(fm.error ?? '', /unterminated/);
});

test('non-mapping frontmatter is an error', () => {
   const fm = parseFrontmatter('---\n- just\n- a list\n---\n\nBody.\n');
   assert.equal(fm.data, null);
   assert.match(fm.error ?? '', /not a YAML mapping/);
});

test('invalid YAML is an error, body still recovered', () => {
   const fm = parseFrontmatter('---\nid: [unclosed\n---\n\nBody.\n');
   assert.equal(fm.data, null);
   assert.match(fm.error ?? '', /invalid YAML/);
   assert.match(fm.body, /Body/);
});

test('YAML 1.2 semantics: unquoted dates and no/yes stay strings', () => {
   const fm = parseFrontmatter('---\ndate: 2026-06-12\nflag: no\nok: true\n---\n\nbody\n');
   assert.equal(fm.data!.date, '2026-06-12');
   assert.equal(fm.data!.flag, 'no');
   assert.equal(fm.data!.ok, true);
});

// --- CRLF line endings ---
// A Windows contributor or core.autocrlf=true authors every line with \r\n. The
// block's last line must keep its whole terminator: a bare \r left behind folds
// into the last scalar's value, which the Go and Python SDKs never did.

/** Join lines with CRLF terminators, the way a Windows editor writes a file. */
const crlf = (...lines: string[]): string => lines.map((line) => `${line}\r\n`).join('');

/** The scan result a profile's bytes produce, as scanFrontmatterArtifact builds it. */
function scanned(relPath: string, text: string): ScannedProfile {
   const fm = parseFrontmatter(text);
   return { relPath, frontmatter: fm.data, body: fm.body, findings: [] };
}

test('CRLF: the last frontmatter field keeps no trailing carriage return', () => {
   const fm = parseFrontmatter(crlf('---', 'id: reviewer', 'role: reviewer', '---', '', 'Body.'));
   assert.equal(fm.error, undefined);
   assert.equal(fm.data!.role, 'reviewer');
   assert.equal(fm.body, '\r\nBody.\r\n');
});

test('CRLF: a profile whose last key is `inherits` passes schema and resolves', () => {
   const base = scanned(
      'docs/agents/core.md',
      crlf(
         '---',
         'id: core',
         'name: Core',
         'role: core',
         'requiredRead:',
         '  - docs/boot-profile.md',
         'mustAskWhen:',
         '  - always',
         '---',
         '',
         '# Core',
      ),
   );
   const derived = scanned(
      'docs/agents/reviewer.md',
      crlf('---', 'id: reviewer', 'name: Reviewer', 'role: reviewer', 'inherits: core', '---', '', '# Reviewer'),
   );
   assert.deepEqual(schemaErrors('agent-profile', derived.frontmatter), []);
   const resolved = resolveAgentProfile(derived, [base, derived]);
   assert.deepEqual(resolved.findings, []);
   assert.deepEqual(resolved.sourceIds, ['core', 'reviewer']);
});

test('CRLF: a blank line before the closing fence still parses', () => {
   const fm = parseFrontmatter(crlf('---', 'id: reviewer', 'role: reviewer', '', '---', '', 'Body.'));
   assert.equal(fm.error, undefined);
   assert.equal(fm.data!.role, 'reviewer');
});

test('CRLF: a decision record reports its schema defect, not a YAML error', () => {
   const fm = parseFrontmatter(
      crlf('---', 'id: crlf-record', 'title: CRLF record', 'date: 2026-07-28', 'status: maybe', '', '---', '', 'Body.'),
   );
   assert.equal(fm.error, undefined);
   const errors = schemaErrors('decision-record', fm.data);
   assert.equal(errors.length, 1);
   assert.match(errors[0], /^\/status /);
});

test('LF documents parse identically, with LF terminators preserved', () => {
   const fm = parseFrontmatter('---\nid: reviewer\nrole: reviewer\n---\n\nBody.\n');
   assert.equal(fm.error, undefined);
   assert.equal(fm.data!.role, 'reviewer');
   assert.equal(fm.body, '\nBody.\n');
});

File: packages/sdk-go/internal/frontmatter/frontmatter_test.go
package frontmatter

import "testing"

func TestNoFrontmatterPassesThrough(t *testing.T) {
	fm := Parse("# Title\n\nBody.\n")
	if fm.Data != nil {
		t.Fatalf("expected nil data, got %v", fm.Data)
	}
	if fm.Error != "" {
		t.Fatalf("expected no error, got %q", fm.Error)
	}
}

func TestUnterminatedBlockIsError(t *testing.T) {
	fm := Parse("---\nid: x\n# never closed\n")
	if fm.Error == "" || !contains(fm.Error, "unterminated") {
		t.Fatalf("expected unterminated error, got %q", fm.Error)
	}
}

func TestNonMappingFrontmatterIsError(t *testing.T) {
	fm := Parse("---\n- just\n- a list\n---\n\nBody.\n")
	if fm.Data != nil {
		t.Fatalf("expected nil data")
	}
	if !contains(fm.Error, "not a YAML mapping") {
		t.Fatalf("expected mapping error, got %q", fm.Error)
	}
}

func TestInvalidYAMLIsErrorBodyRecovered(t *testing.T) {
	fm := Parse("---\nid: [unclosed\n---\n\nBody.\n")
	if fm.Data != nil {
		t.Fatalf("expected nil data")
	}
	if !contains(fm.Error, "invalid YAML") {
		t.Fatalf("expected invalid YAML, got %q", fm.Error)
	}
	if !contains(fm.Body, "Body") {
		t.Fatalf("body not recovered: %q", fm.Body)
	}
}

func TestYAML12Semantics(t *testing.T) {
	fm := Parse("---\ndate: 2026-06-12\nflag: no\nok: true\n---\n\nbody\n")
	if fm.Data == nil {
		t.Fatal("expected data")
	}
	if d, ok := fm.Data["date"].(string); !ok || d != "2026-06-12" {
		t.Fatalf("date should stay string, got %#v", fm.Data["date"])
	}
	if f, ok := fm.Data["flag"].(string); !ok || f != "no" {
		t.Fatalf("flag 'no' should stay string, got %#v", fm.Data["flag"])
	}
	if b, ok := fm.Data["ok"].(bool); !ok || !b {
		t.Fatalf("ok 'true' should be bool true, got %#v", fm.Data["ok"])
	}
}

// Parse slices the raw YAML block to the end of the fence's submatch 1, so that
// submatch must open at the match start and span the whole line terminator ending
// the block's last line. This is asserted on the regex rather than through Parse
// because yaml.v3 normalizes a bare trailing `\r`: no Parse-level assertion can
// fail when this regresses. The Node SDK's parser does not normalize it, and folds
// the orphan CR into the last scalar's value, which is what the shared slicing
// protects against here and in packages/sdk-py.
func TestFenceSubmatchSpansTheWholeLineTerminator(t *testing.T) {
	cases := []struct {
		name  string
		block string
		want  string
	}{
		{"crlf", "\r\nrole: reviewer\r\n---\r\n", "\r\n"},
		{"lf", "\nrole: reviewer\n---\n", "\n"},
	}
	for _, tc := range cases {
		loc := fence.FindStringSubmatchIndex(tc.block)
		if loc == nil {
			t.Fatalf("%s: fence did not match %q", tc.name, tc.block)
		}
		if loc[2] != loc[0] {
			t.Fatalf("%s: submatch 1 must open at the match start; Parse slices to its end", tc.name)
		}
		if got := tc.block[loc[2]:loc[3]]; got != tc.want {
			t.Fatalf("%s: terminator submatch = %q, want %q", tc.name, got, tc.want)
		}
	}
}

func TestDuplicateKeyIsError(t *testing.T) {
	fm := Parse("---\nid: a\nid: b\n---\n\nbody\n")
	if fm.Error == "" {
		t.Fatalf("expected error for duplicate key, got data %v", fm.Data)
	}
}

func contains(s, sub string) bool {
	return len(s) >= len(sub) && (indexOf(s, sub) >= 0)
}

func indexOf(s, sub string) int {
	for i := 0; i+len(sub) <= len(s); i++ {
		if s[i:i+len(sub)] == sub {
			return i
		}
	}
	return -1
}
